File: iOS-SDK/PhoneAgentDemo/SDK/Model/ModelConfig.swift
//
//  ModelConfig.swift
//  PhoneAgentSDK
//
//  Created by AutoGLM on 2025-01-30.
//

import Foundation

/// 模型配置
public struct ModelConfig {

    /// API 基础地址
    public var baseURL: String

    /// API 密钥
    public var apiKey: String

    /// 模型名称
    public var modelName: String

    /// 最大生成 token 数
    public var maxTokens: Int

    /// 温度参数
    public var temperature: Double

    /// Top P 采样
    public var topP: Double

    /// 频率惩罚
    public var frequencyPenalty: Double

    /// 语言（用于提示词和消息）
    public var language: Language

    public enum Language: String {
        case chinese = "cn"
        case english = "en"

        var localeCode: String {
            switch self {
            case .chinese: return "zh-CN"
            case .english: return "en-US"
            }
        }
    }

    public init(
        baseURL: String = "http://localhost:8000/v1",
        apiKey: String = "EMPTY",
        modelName: String = "autoglm-phone-9b",
        maxTokens: Int = 3000,
        temperature: Double = 0.0,
        topP: Double = 0.85,
        frequencyPenalty: Double = 0.2,
        language: Language = .chinese
    ) {
        self.baseURL = baseURL
        self.apiKey = apiKey
        self.modelName = modelName
        self.maxTokens = maxTokens
        self.temperature = temperature
        self.topP = topP
        self.frequencyPenalty = frequencyPenalty
        self.language = language
    }
}

// MARK: - 默认配置
extension ModelConfig {
    /// 默认配置（中文）
    public static let `default` = ModelConfig()

    /// 英文配置
    public static let english = ModelConfig(language: .english)
}

File: Android-SDK/phoneagent-sdk/src/main/java/com/autoglm/phoneagent/model/ModelConfig.kt
package com.autoglm.phoneagent.model

/**
 * Configuration for the AutoGLM model client.
 *
 * @property baseURL Base URL for the API endpoint
 * @property apiKey API key for authentication
 * @property modelName Name of the model to use
 * @property maxTokens Maximum tokens in the response (default: 3000)
 * @property temperature Sampling temperature (default: 0.0)
 * @property topP Top-p sampling parameter (default: 0.85)
 * @property frequencyPenalty Frequency penalty (default: 0.2)
 * @property language Language for prompts (Chinese or English)
 */
data class ModelConfig(
    val baseURL: String = "https://open.bigmodel.cn/api/paas/v4",
    val apiKey: String = "",
    val modelName: String = "autoglm-phone",
    val maxTokens: Int = 3000,
    val temperature: Double = 0.0,
    val topP: Double = 0.85,
    val frequencyPenalty: Double = 0.2,
    val language: Language = Language.CHINESE
) {
    /**
     * Clean and normalize the base URL to fix common input errors.
     * - Removes accidental leading 'c' (e.g., 'chttps' -> 'https')
     * - Ensures proper URL format
     */
    private fun cleanURL(url: String): String {
        var cleaned = url.trim()

        // Fix common typo: chttps -> https
        if (cleaned.startsWith("chttps")) {
            cleaned = "https" + cleaned.removePrefix("chttps")
        } else if (cleaned.startsWith("chttp")) {
            cleaned = "http" + cleaned.removePrefix("chttp")
        }

        return cleaned
    }

    /**
     * Get the cleaned base URL.
     */
    val cleanBaseURL: String
        get() = cleanURL(baseURL)
    /**
     * Supported languages for system prompts.
     */
    enum class Language(val code: String) {
        CHINESE("cn"),
        ENGLISH("en");

        companion object {
            fun fromCode(code: String): Language {
                return values().find { it.code == code } ?: CHINESE
            }
        }
    }

    companion object {
        /**
         * Default configuration for Chinese language.
         */
        val DEFAULT = ModelConfig()

        /**
         * Default configuration for English language.
         */
        val ENGLISH = ModelConfig(language = Language.ENGLISH)

        /**
         * Local development configuration.
         */
        val LOCAL = ModelConfig(
            baseURL = "http://localhost:8000/v1",
            modelName = "autoglm-phone-9b"
        )

        /**
         * ModelScope configuration.
         */
        val MODELSCOPE = ModelConfig(
            baseURL = "https://api-inference.modelscope.cn/v1",
            modelName = "autoglm-phone-9b"
        )
    }

    /**
     * Validate the configuration.
     * @throws IllegalArgumentException if configuration is invalid
     */
    fun validate() {
        val url = cleanBaseURL
        require(url.isNotBlank()) { "baseURL must not be blank" }
        require(apiKey.isNotBlank()) { "apiKey must not be blank" }
        require(modelName.isNotBlank()) { "modelName must not be blank" }
        require(maxTokens > 0) { "maxTokens must be greater than 0" }
        require(temperature in 0.0..2.0) { "temperature must be between 0.0 and 2.0" }
        require(topP in 0.0..1.0) { "topP must be between 0.0 and 1.0" }
        require(frequencyPenalty in -2.0..2.0) { "frequencyPenalty must be between -2.0 and 2.0" }
    }

    /**
     * Get the full chat completion URL.
     */
    fun getChatCompletionUrl(): String {
        val url = cleanBaseURL
        return when {
            url.endsWith("/chat/completions") -> url
            url.endsWith("/v1") -> "$url/chat/completions"
            url.endsWith("/paas/v4") -> "$url/chat/completions"
            else -> "$url/chat/completions"
        }
    }
}
